streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -374,19 +373,30 @@ StreamTask createTask(final Consumer<byte[], byte[]> consumer,
                 storeChangelogReader,
                 logContext);
 
+            if (threadProducer == null) {
+                // create one producer per task for EOS
+                // TODO: after KIP-447 this would be removed
+                final Map<String, Object> producerConfigs = config.getProducerConfigs(getTaskProducerClientId(threadId, taskId));
+                producerConfigs.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, applicationId + "-" + taskId);
+                log.info("Creating producer client for task {}", taskId);
+                taskProducer.put(taskId, clientSupplier.getProducer(producerConfigs));
```

Comment:
When are we removing the entry upon task closure? If it never cleans up we could potentially have an ever-growing map.

Reply:
The `TaskManager` does this in `handleAssignment()` and `handleLostAll()`.